tsdb/store_test.go

```diff
@@ -262,6 +262,74 @@ func TestStore_BadShard(t *testing.T) {
 	}
 }
 
+func TestStore_BadShardClear(t *testing.T) {
+	const errStr = "a shard open error"
+	indexes := tsdb.RegisteredIndexes()
+	for _, idx := range indexes {
+		func() {
+			s := MustOpenStore(t, idx)
+			defer require.NoErrorf(t, s.Close(), "closing store with index type: %s", idx)
+
+			sh := tsdb.NewTempShard(idx)
+			shId := sh.ID()
+			err := s.OpenShard(sh.Shard, false)
+			require.NoError(t, err, "opening temp shard")
+			defer require.NoError(t, sh.Close(), "closing temporary shard")
+
+			expErr := errors.New(errStr)
+			s.SetShardOpenErrorForTest(sh.ID(), expErr)
+			err2 := s.OpenShard(sh.Shard, false)
+			require.Error(t, err2, "no error opening bad shard")
+			require.True(t, errors.Is(err2, tsdb.ErrPreviousShardFail{}), "exp: ErrPreviousShardFail, got: %v", err2)
+			require.EqualError(t, err2, fmt.Errorf("not attempting to open shard %d; opening shard previously failed with: %w", shId, expErr).Error())
+
+			require.Equal(t, 1, len(s.Store.GetBadShardList()))
+
+			badShards := s.ClearBadShardList()
+			if len(badShards) != 1 {
+				t.Fatalf("expected 1 shard, got %d", len(badShards))
+			}
+
+			// Check that bad shard list has been cleared
+			require.Equal(t, 0, len(s.Store.GetBadShardList()))
+
+			s.SetShardOpenErrorForTest(sh.ID(), expErr)
+			err2 = s.OpenShard(sh.Shard, false)
+			require.Error(t, err2, "no error opening bad shard")
+			require.True(t, errors.Is(err2, tsdb.ErrPreviousShardFail{}), "exp: ErrPreviousShardFail, got: %v", err2)
+			require.EqualError(t, err2, fmt.Errorf("not attempting to open shard %d; opening shard previously failed with: %w", shId, expErr).Error())
+
+			// Check that bad shard list now has a bad shard in it
+			require.Equal(t, 1, len(s.Store.GetBadShardList()))
+		}()
+	}
+}
+
+func TestStore_BadShardClearNoBadShards(t *testing.T) {
+	indexes := tsdb.RegisteredIndexes()
+	for _, idx := range indexes {
+		func() {
+			s := MustOpenStore(t, idx)
+			defer require.NoErrorf(t, s.Close(), "closing store with index type: %s", idx)
```

Comment:
Nice error checking!

Reply:
Isn't `s.Close` called on the `defer` line?

> 1. A deferred function’s arguments are evaluated when the defer statement is evaluated.

https://go.dev/blog/defer-panic-and-recover

Reply:
> Isn't `s.Close` called on the `defer` line?
> 
> > 1. A deferred function’s arguments are evaluated when the defer statement is evaluated.
> 
> [go.dev/blog/defer-panic-and-recover](https://go.dev/blog/defer-panic-and-recover)

Is this the expectation? I assumed that `s.Close()` would be called as per the defer so the end of scope. Which is expected in the test.